pyirf/interpolation.py

```diff
@@ -0,0 +1,190 @@
+        grid of effective area, of shape (n_grid_points, n_fov_offset_bins, n_energy_bins)
+    grid_points: np.array
+        list of parameters corresponding to effective_area, of shape (n_grid_points, n_interp_dim)
+    target_point: np.array
+        values of parameters for which the interpolation is performed, of shape (n_interp_dim)
+    min_effective_area: astropy.units.Quantity[area]
+        Minimum value of effective area to be considered for interpolation
+    method: 'linear’, ‘nearest’, ‘cubic’
+        Interpolation method
+
+    Returns
+    -------
+    aeff_interp: astropy.units.Quantity[area]
+        Interpolated Effective area array with shape (n_energy_bins, n_fov_offset_bins)
+    """
+
+    _, n_fov_offset_bins, n_energy_bins = effective_area.shape
+
+    # get rid of units
+    effective_area = effective_area.to_value(u.m**2)
+    min_effective_area = min_effective_area.to('m2').value
+
+    # remove zeros and log it
+    effective_area[effective_area < min_effective_area] = min_effective_area
+    effective_area = np.log(effective_area)
+
+    # interpolation
+    aeff_interp = griddata(grid_points, effective_area, target_point, method=method).T
+    # exp it and set to zero too low values
+    aeff_interp = np.exp(aeff_interp)
+    aeff_interp[aeff_interp < min_effective_area * 1.1] = 0  # 1.1 to correct for numerical uncertainty and interpolation
+    return u.Quantity(aeff_interp, u.m**2, copy=False)
+
+
+def interpolate_energy_dispersion(energy_dispersions, grid_points, target_point, method='linear'):
+    """
+    Takes a grid of dispersion matrixes for a bunch of different parameters
+    and interpolates it to given value of those parameters
+
+    Parameters
+    ----------
+    energy_dispersions: np.array of astropy.units.Quantity[area]
+        grid of effective area, of shape (n_grid_points, n_fov_offset_bins, n_migration_bins, n_energy_bins)
+    grid_points: np.array
+        list of parameters corresponding to energy_dispersions, of shape (n_grid_points, n_interp_dim)
+    target_point: np.array
+        values of parameters for which the interpolation is performed, of shape (n_interp_dim)
+    method: 'linear’, ‘nearest’, ‘cubic’
+        Interpolation method
+
+    Returns
+    -------
+    matrix_interp: astropy.units.Quantity[area]
+        Interpolated dispersion matrix 3D array with shape (n_energy_bins, n_migration_bins, n_fov_offset_bins)
+    """
+
+    _, n_fov_offset_bins, n_migration_bins, n_energy_bins = energy_dispersions.shape
+
+    # interpolation
+    matrix_interp = griddata(grid_points, energy_dispersions, target_point, method=method)
+    matrix_interp = np.swapaxes(matrix_interp, 0, 2)
+
+    # now we need to renormalize along the migration axis
+    norm = np.sum(matrix_interp, axis=1, keepdims=True)
+    mig_norm = np.divide(matrix_interp, norm, out=np.zeros_like(matrix_interp), where=norm != 0)
+    return mig_norm
+
+
+def read_fits_bins_lo_hi(file_name, ext_name, tag):
```

Comment:
If you want to implement these reading functions, Please move them to `io/gadf.py` and have one reading method per IRF component that returns all the relevant information from the HDU. 

Not one function to read the bins, one function to read the irf values and so on.

Also, again you are mixing two things, the actual IO of reading in the data and trasforming it (combining lo + hi to contiguous bins).

This is not ok.

The function to combine lo / hi to edges assuming contiguous bins (which we right now use in pyirf but are not mandated by the standard!) can go into `pyirf/binning.py`.

Reply:
> If you want to implement these reading functions, Please move them to `io/gadf.py` 

ok, will do

> and have one reading method per IRF component that returns all the relevant information from the HDU.
> 
> Not one function to read the bins, one function to read the irf values and so on.
> 

note that the current scheme uses a single function for reading any of the different IRFs  (tested on effective area and migration matrix, but should work also fine for PSF) and the second function to read the binning, so it is not 2 functions per IRF type, but 2 functions in total (in contrary to what you propose with 1 function per IRF type), but no problem I can make wrap ups for individual IRFs that would extract the values and binning (and specify the default extension and field names for the variables passed to those functions)

> Also, again you are mixing two things, the actual IO of reading in the data and trasforming it (combining lo + hi to contiguous bins).
> 
> This is not ok.
> 
> The function to combine lo / hi to edges assuming contiguous bins (which we right now use in pyirf but are not mandated by the standard!) can go into `pyirf/binning.py`.

the transformation is from the two lists of bin edges (lo and hi) to a single list. It can be extracted to a separate function, but the same kind of transformation (but the other way around) is also used in all the functions of gadf.py: create_aeff2d_hdu, create_psf_table_hdu, create_energy_dispersion_hdu, create_background_2d_hdu, create_rad_max_hdu. 
So we should then also extract it from those functions (which might be a good idea actually, as it will avoid quite some code duplication)

Reply:
> (which might be a good idea actually, as it will avoid quite some code duplication)

Yes!

Reply:
the functions reading the binning from FITS file and joining/splitting it are now separated